src/Cassandra/Requests/RequestExecution.cs

```diff
@@ -501,47 +501,66 @@ private void PrepareAndRetry(byte[] id, Host host)
                     .SetKeyspace(preparedKeyspace)
                     .ContinueSync(_ =>
                     {
-                        Send(request, host, ReprepareResponseHandler);
+                        Send(request, host, NewReprepareResponseHandler(ex));
                         return true;
                     });
                 return;
             }
-            Send(request, host, ReprepareResponseHandler);
+            Send(request, host, NewReprepareResponseHandler(ex));
         }
 
         /// <summary>
         /// Handles the response of a (re)prepare request and retries to execute on the same connection
         /// </summary>
-        private void ReprepareResponseHandler(IRequestError error, Response response, Host host)
+        private Action<IRequestError, Response, Host> NewReprepareResponseHandler(
+            PreparedQueryNotFoundException originalError)
         {
-            try
+            void ResponseHandler(IRequestError error, Response response, Host host)
             {
-                if (error?.Exception != null)
-                {
-                    HandleRequestError(error, host);
-                    return;
-                }
-                RequestExecution.ValidateResult(response);
-                var output = ((ResultResponse)response).Output;
-                if (!(output is OutputPrepared outputPrepared))
+                try
                 {
-                    _parent.SetCompleted(
-                        new DriverInternalError("Expected prepared response, obtained " + output.GetType().FullName));
-                    return;
-                }
+                    if (error?.Exception != null)
+                    {
+                        HandleRequestError(error, host);
+                        return;
+                    }
+
+                    RequestExecution.ValidateResult(response);
+                    var output = ((ResultResponse) response).Output;
+                    if (!(output is OutputPrepared outputPrepared))
+                    {
+                        _parent.SetCompleted(new DriverInternalError("Expected prepared response, obtained " + output.GetType().FullName));
+                        return;
+                    }
+
+                    if (!outputPrepared.QueryId.SequenceEqual(originalError.UnknownId))
+                    {
+                        throw new PreparedStatementIdMismatchException(
```

Comment:
@jorgebay now that I think about it, we should probably replace this with 

```
_parent.SetCompleted(new PreparedStatementIdMismatchException(...));
return;
```

This has performance benefits and is more consistent with the way we are returning errors in this same method.

Reply:
The downside is that we lose stacktrace information (I think)

Reply:
oh, good catch, yes I think it should be done with `SetCompleted()` (!)